src/sources/http_scrape/scrape.rs

```diff
@@ -0,0 +1,243 @@
+//! Generalized HTTP scrape source.
+//! Scrapes an endpoint at an interval, decoding the HTTP responses into events.
+
+use bytes::{Bytes, BytesMut};
+use chrono::Utc;
+use futures_util::FutureExt;
+use http::{response::Parts, Uri};
+use snafu::ResultExt;
+use std::collections::HashMap;
+use tokio_util::codec::Decoder as _;
+
+use crate::{
+    codecs::{Decoder, DecodingConfig},
+    config::{SourceConfig, SourceContext, SourceDescription},
+    http::Auth,
+    serde::default_decoding,
+    serde::default_framing_message_based,
+    sources,
+    tls::{TlsConfig, TlsSettings},
+    Result,
+};
+use codecs::{
+    decoding::{DeserializerConfig, FramingConfig},
+    StreamDecodingError,
+};
+use vector_config::configurable_component;
+use vector_core::{
+    config::{log_schema, LogNamespace, Output},
+    event::Event,
+};
+
+/// The name of this source
+pub(crate) const NAME: &str = "http_scrape";
+
+/// Configuration for the `http_scrape` source.
+#[configurable_component(source)]
+#[derive(Clone, Debug)]
+pub struct HttpScrapeConfig {
+    /// Endpoint to scrape events from. The full path must be specified.
+    /// Example: "http://127.0.0.1:9898/logs"
+    endpoint: String,
+
+    /// The interval between scrapes, in seconds.
+    #[serde(default = "super::default_scrape_interval_secs")]
+    scrape_interval_secs: u64,
+
+    /// Custom parameters for the scrape request query string.
+    ///
+    /// One or more values for the same parameter key can be provided. The parameters provided in this option are
+    /// appended to any parameters manually provided in the `endpoint` option.
+    query: Option<HashMap<String, Vec<String>>>,
+
+    /// Decoder to use on the HTTP responses.
+    #[configurable(derived)]
+    #[serde(default = "default_decoding")]
+    decoding: DeserializerConfig,
+
+    /// Framing to use in the decoding.
+    #[configurable(derived)]
+    #[serde(default = "default_framing_message_based")]
+    framing: FramingConfig,
+
+    /// Headers to apply to the HTTP requests.
+    /// One or more values for the same header can be provided.
+    #[serde(default)]
+    headers: Option<HashMap<String, Vec<String>>>,
+
+    /// TLS configuration.
+    #[configurable(derived)]
+    tls: Option<TlsConfig>,
+
+    /// HTTP Authentication.
+    #[configurable(derived)]
+    auth: Option<Auth>,
+}
+
+impl Default for HttpScrapeConfig {
```

Comment:
Do we use this default implementation anywhere? If not, it might be better to just add the default `endpoint` to the struct above and be done.

Reply:
The default impl is required by `impl_generate_config_from_default!` , which I discovered looking at some other sources. I thought it looked a little cleaner and more portable than implementing the generate explicitly in this file.